decisions/0003-cf-api-controller-repo.md

```diff
@@ -0,0 +1,89 @@
+# 3. CF API Controller Source Code Organization
+
+Date: 2020-06-17
+
+## Status
+
+In Review
+
+## Context
+
+Currently we store `capi-kpack-watcher` directly in `capi-k8s-release`. This has
```

Comment:
I'd love to understand this pain point a little more. Have we reached out to other teams like cf-k8s-networking; their release resembles ours a lot, and they're using kubebuilder too!

For me personally, I have not seen the docker image as a weird reference to capi-k8s-release...but to an asset that happens to be sourced in the same repo. Bumping reflects a change set of of consuming a new docker image for a component (and maybe any configuration changes that come with it). Feels super analogous to bosh releases.

I'm not yet sold on the idea on the idea of moving the code out of capi-k8s-release. Maybe hearing from someone on Networking would help.

Reply:
I think the strangeness is how the "bumped asset" likely doesn't match the checked-in code half the time. We update the capi-k8s-release source, it gets rebuilt, we deploy it, and then if everything works, we make another commit that "bumps" the image. This can cause red "bump" jobs when we try to put those changes on top of as-of-yet unbuilt changes pushed to master. 

I 100% agree we should talk to the networking folks about how their release structure is working for them. The more we talk about this the more distant the pipelining where we thought this was a huge problem seems to me.

Reply:
> I think the strangeness is how the "bumped asset" likely doesn't match the checked-in code half the time.

IMO as long as the asset matches the code at the time the repo is tagged and a release is cut, it isn't that confusing.

We typically don't have frequent commits or multiple-pairs working on `routecontroller` at the same time so the brief lag time between the commit, tests to run, and the bump hasn't caused problems. This is our pipeline that does that stuff, by the way:
https://networking.ci.cf-app.com/teams/cf-k8s/pipelines/cf-k8s-pipeline?group=routecontroller

Reply:
I do think this has forced us to do odd things in Concourse. Our pipeline doesn't feel like one continuous pipeline, we have a group for our routecontroller that watches the routecontroller directory and runs the unit/integration tests and then bumps the image in our `config/values.yaml`. Then another part of our pipeline watches for changes to the config directory and runs our acceptance tests. It basically runs as this contiguous thing, but looks broken up.

That being said breaking it up into repos, probably still forces you to do the same thing, but maybe its clearer that they are separate.

Reply:
I'm still orienting myself to how we structure code in the k8s world, so I don't have a strong opinion at the moment. That said, it doesn't sound to me like the current situation is really all that bad or that the alternatives would be that large of an improvement. Given that, I have a bias for inaction.

Reply:
@christianang Our pipeline works slightly differently... We build the OCI images after unit tests, but we don't actually bump `config/values/images.yml` until after integration tests have passed.